File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val   int
 *     Left  *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(preorder []int, inorder []int) *TreeNode {
    // preIndex is the index of the current value in preorder.
    preIndex := 0
    
    // valToIndex maps the values in inorder to their index.
    // This saves us from an O(N) search of inorder each iteration.
    valToIndex := make(map[int]int)
    for i, v := range inorder {
        valToIndex[v] = i
    }
    
    var build func(head, tail int) *TreeNode
    
    build = func(head, tail int) *TreeNode {
        if (head > tail) {
            return nil
        }
        
        rootVal := preorder[preIndex]
        rootIndex := valToIndex[rootVal]
        preIndex += 1

        return &TreeNode{
            Val:   rootVal,
            Left:  build(head, rootIndex - 1),
            Right: build(rootIndex + 1, tail),
        }
    }
    
    return build(0, len(preorder) - 1)
}

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        pre_ind = 0
        
        val_to_ind = {v: i for i, v in enumerate(inorder)}
        
        def build(head: int, tail: int) -> Optional[TreeNode]:
            if head > tail:
                return None
            
            nonlocal pre_ind
            
            root_val = preorder[pre_ind]
            root_ind = val_to_ind[root_val]
            pre_ind += 1
            
            return TreeNode(
                root_val,
                build(head, root_ind - 1),
                build(root_ind + 1, tail)
            )
        
        return build(0, len(preorder) - 1)